### PR TITLE
Scoped queries

### LINKED ISSUE #51: Forbid `trusting previous` in the authorizer block

### PR DESCRIPTION
**Pull request by divarvel**
Closes #51, #52

- `queryAuthorizerFacts` now provides access to the complete set of facts generated during authorizaton.
The default behaviour stays the same: un-scoped queries only give access to authority & authorizer facts.
It is now possible to add a scope to the query (using `trusting …`) to access facts from all blocks.

- `queryRawBiscuitFacts` now provides a way to query facts declared in a biscuit (facts derived through rules are not queried).
  Same as for `queryAuthorizerFacts`, query scopes can be used to extract facts from various blocks.
  This function is made to be used when generating the authorizer requires information from the token. The canonical case would be extracting the user id from the token to load authorization rules pertaining to this user only, instead of loading the complete rights matrix

- `trusting previous` is now forbidden in the authorizer, in accordance to the rust implementation and the spec. `trusting previous` is still allowed in `queryAuthorizerFacts`.

### CODE REVIEW

File: biscuit/src/Auth/Biscuit/Datalog/Executor.hs [hunk 1]
```diff
@@ -158,13 +158,17 @@ keepAuthorized (FactGroup facts) authorizedOrigins =
   let isAuthorized k _ = k `Set.isSubsetOf` authorizedOrigins
    in FactGroup $ Map.filterWithKey isAuthorized facts
 
-keepAuthorized' :: Natural -> FactGroup -> Set EvalRuleScope -> Natural -> FactGroup
-keepAuthorized' blockCount factGroup trustedBlocks currentBlockId =
+keepAuthorized' :: Bool -> Natural -> FactGroup -> Set EvalRuleScope -> Natural -> FactGroup
```
Comment (ptitfred) [inline biscuit/src/Auth/Biscuit/Datalog/Executor.hs:161]: Not concerned by boolean blindness here?

File: biscuit/src/Auth/Biscuit/Datalog/Executor.hs [hunk 2]
```diff
@@ -158,13 +158,17 @@ keepAuthorized (FactGroup facts) authorizedOrigins =
...
+keepAuthorized' allowPreviousInAuthorizer blockCount factGroup trustedBlocks currentBlockId =
   let scope = if null trustedBlocks then Set.singleton OnlyAuthority
                                     else trustedBlocks
       toBlockIds = \case
         OnlyAuthority    -> Set.singleton 0
-        Previous         -> Set.fromList [0..currentBlockId]
+        Previous         -> if allowPreviousInAuthorizer || currentBlockId < blockCount
```
Comment (ptitfred) [inline biscuit/src/Auth/Biscuit/Datalog/Executor.hs:167]: Is arithmetic on block ids safe?

File: biscuit/src/Auth/Biscuit/Datalog/Executor.hs [hunk 3]
```diff
@@ -158,13 +158,17 @@ keepAuthorized (FactGroup facts) authorizedOrigins =
...
+                            then Set.fromList [0..currentBlockId]
+                            else mempty -- `Previous` is forbidden in the authorizer
```
Comment (ptitfred) [inline biscuit/src/Auth/Biscuit/Datalog/Executor.hs:169]: Would the user benefit from any form of warning/hint in such a case?

File: biscuit/src/Auth/Biscuit/Datalog/Parser.hs [hunk 4]
```diff
@@ -54,16 +54,18 @@ data SemanticError =
...
-    InvalidBs e _        -> "Invalid bytestring literal: " <> T.unpack e
-    InvalidPublicKey e _ -> "Invalid public key: " <> T.unpack e
-    UnboundVariables e _ -> "Unbound variables: " <> T.unpack (T.intercalate ", " $ NE.toList e)
+    VarInFact _            -> "Variables can't appear in a fact"
+    VarInSet  _            -> "Variables can't appear in a set"
+    NestedSet _            -> "Sets cannot be nested"
+    InvalidBs e _          -> "Invalid bytestring literal: " <> T.unpack e
+    InvalidPublicKey e _   -> "Invalid public key: " <> T.unpack e
+    UnboundVariables e _   -> "Unbound variables: " <> T.unpack (T.intercalate ", " $ NE.toList e)
+    PreviousInAuthorizer _ -> "'previous' can't appear in an authorizer scope"
```
Comment (ptitfred) [inline biscuit/src/Auth/Biscuit/Datalog/Parser.hs:68]: Maybe this covers my previous comment regarding warning the user.

File: biscuit/src/Auth/Biscuit/Datalog/ScopedExecutor.hs [hunk 5]
```diff
@@ -276,23 +278,18 @@ collectWorld blockId Block{..} =
       , FactGroup $ Map.singleton (Set.singleton blockId) $ Set.fromList bFacts
       )
 
--- | Query the facts generated by the authority and authorizer blocks
```
Comment (ptitfred) [inline biscuit/src/Auth/Biscuit/Datalog/ScopedExecutor.hs:279]: Looks like you dropped a function and its documentation to replace it by two new ones without haddock. Was it intended?

File: biscuit/src/Auth/Biscuit/Token.hs [hunk 6]
```diff
@@ -314,6 +357,12 @@ applyThirdPartyBlock b@Biscuit{..} contents = do
              , symbols = newSymbols
              }
 
+externalKeys :: Biscuit openOrSealed check -> [Maybe PublicKey]
```
Comment (ptitfred) [inline biscuit/src/Auth/Biscuit/Token.hs:360]: Is external an alias to third party?

**Review by ptitfred — APPROVED** [review]

File: biscuit/src/Auth/Biscuit/Datalog/ScopedExecutor.hs (hunk 5 above)
Comment (divarvel) [reply biscuit/src/Auth/Biscuit/Datalog/ScopedExecutor.hs:279]: The function has been moved to `Biscuit.Token`, where it's properly documented, along with the new `queryRawBiscuitFacts`

The two new functions in this module are not meant to be used directly.

File: biscuit/src/Auth/Biscuit/Token.hs (hunk 6 above)
Comment (divarvel) [reply biscuit/src/Auth/Biscuit/Token.hs:360]: Yes.

File: biscuit/src/Auth/Biscuit/Datalog/Executor.hs (hunk 2 above)
Comment (divarvel) [reply biscuit/src/Auth/Biscuit/Datalog/Executor.hs:167]: Yes: the block id is the index in the block list, so range operations exactly match the semantics of `Previous`

**Comment by ptitfred** [comment]
Does it fixes #51 ?

**Comment by ptitfred** [comment]
And maybe #52 too?

**Comment by divarvel** [comment]
Yes, it fixes both.

File: biscuit/src/Auth/Biscuit/Datalog/Executor.hs (hunk 1 above)
Comment (divarvel) [reply biscuit/src/Auth/Biscuit/Datalog/Executor.hs:161]: Just a bit, as always, but in this case it's fairly contained and not exposed to outside use

**Pull request merged**